File: sdk/go/ancora/xlang_single_agent_test.go
package ancora_test

import (
	"encoding/json"
	"testing"
)

// Cross-language conformance: single agent scenario -- Go binding.
// Offline: local fake runtime with a fixed event queue.

type xlangFakeRuntime struct {
	runID  string
	events [][]byte
	pos    int
}

func (f *xlangFakeRuntime) poll() []byte {
	if f.pos >= len(f.events) {
		return nil
	}
	ev := f.events[f.pos]
	f.pos++
	return ev
}

func newXlangFakeRuntime(runID string) *xlangFakeRuntime {
	events := [][]byte{
		[]byte(`{"kind":"started","run_id":"` + runID + `","spec":"{}"}`),
		[]byte(`{"kind":"token","run_id":"` + runID + `","text":"xlang go result"}`),
		[]byte(`{"kind":"completed","run_id":"` + runID + `"}`),
	}
	return &xlangFakeRuntime{runID: runID, events: events}
}

type xlangGoEvent struct {
	Kind  string `json:"kind"`
	RunID string `json:"run_id"`
	Text  string `json:"text,omitempty"`
}

func drainXlang(fr *xlangFakeRuntime) []xlangGoEvent {
	var out []xlangGoEvent
	for {
		raw := fr.poll()
		if raw == nil {
			break
		}
		var ev xlangGoEvent
		json.Unmarshal(raw, &ev) //nolint:errcheck
		out = append(out, ev)
	}
	return out
}

func TestXlangGoSingleAgentStartedEventFirst(t *testing.T) {
	fr := newXlangFakeRuntime("xlang-go-001")
	events := drainXlang(fr)
	if len(events) == 0 {
		t.Fatal("expected events")
	}
	if events[0].Kind != "started" {
		t.Fatalf("first event must be started, got %q", events[0].Kind)
	}
}

func TestXlangGoSingleAgentRunIDConsistent(t *testing.T) {
	const rid = "xlang-go-002"
	fr := newXlangFakeRuntime(rid)
	for _, ev := range drainXlang(fr) {
		if ev.RunID != rid {
			t.Fatalf("run_id mismatch: want %q got %q", rid, ev.RunID)
		}
	}
}

func TestXlangGoSingleAgentCompletedEventLast(t *testing.T) {
	fr := newXlangFakeRuntime("xlang-go-003")
	events := drainXlang(fr)
	last := events[len(events)-1]
	if last.Kind != "completed" {
		t.Fatalf("last event must be completed, got %q", last.Kind)
	}
}

func TestXlangGoSingleAgentEventCount(t *testing.T) {
	fr := newXlangFakeRuntime("xlang-go-004")
	events := drainXlang(fr)
	if len(events) < 2 {
		t.Fatalf("expected at least 2 events, got %d", len(events))
	}
}

func TestXlangGoSingleAgentTokenTextNonEmpty(t *testing.T) {
	fr := newXlangFakeRuntime("xlang-go-005")
	var found bool
	for _, ev := range drainXlang(fr) {
		if ev.Kind == "token" && ev.Text != "" {
			found = true
		}
	}
	if !found {
		t.Fatal("expected at least one non-empty token event")
	}
}

File: sdk/ts/__tests__/xlang-single-agent.test.ts
// Cross-language conformance: single agent scenario -- TypeScript binding (offline).

interface XlangEvent {
  kind: string
  run_id: string
  text?: string
  spec?: string
}

const XLANG_RUN_ID = 'xlang-ts-001'

function makeXlangEvents(runId: string): XlangEvent[] {
  return [
    { kind: 'started', run_id: runId, spec: '{}' },
    { kind: 'token', run_id: runId, text: 'xlang ts result' },
    { kind: 'completed', run_id: runId },
  ]
}

describe('xlang single agent -- TypeScript', () => {
  const events = makeXlangEvents(XLANG_RUN_ID)

  it('started event is first', () => {
    expect(events[0].kind).toBe('started')
  })

  it('completed event is last', () => {
    expect(events[events.length - 1].kind).toBe('completed')
  })

  it('run_id is consistent across all events', () => {
    for (const ev of events) {
      expect(ev.run_id).toBe(XLANG_RUN_ID)
    }
  })

  it('has at least two events', () => {
    expect(events.length).toBeGreaterThanOrEqual(2)
  })

  it('token event has non-empty text', () => {
    const tokens = events.filter(e => e.kind === 'token')
    expect(tokens.length).toBeGreaterThan(0)
    for (const tok of tokens) {
      expect(tok.text).toBeTruthy()
    }
  })

  it('events serialise and deserialise cleanly', () => {
    for (const ev of events) {
      const serialised = JSON.stringify(ev)
      const decoded = JSON.parse(serialised) as XlangEvent
      expect(decoded.kind).toBe(ev.kind)
    }
  })

  it('first event spec field is valid JSON', () => {
    const first = events[0]
    expect(first.kind).toBe('started')
    const parsed = JSON.parse(first.spec ?? '{}')
    expect(typeof parsed).toBe('object')
  })
})
